src/Microsoft.Diagnostics.Monitoring.WebApi/Metrics/StreamingCounterLogger.cs

```diff
@@ -3,39 +3,71 @@
 // See the LICENSE file in the project root for more information.
 
 using Microsoft.Diagnostics.Monitoring.EventPipe;
-using System.IO;
+using System;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
 
 namespace Microsoft.Diagnostics.Monitoring.WebApi
 {
     internal abstract class StreamingCounterLogger : ICountersLogger
     {
-        private readonly Stream _stream;
+        private const int BoundedChannelSize = 1000;
 
-        protected abstract void SerializeCounter(Stream stream, ICounterPayload counter);
+        // The amount of time to wait for serialization to finish before stopping the pipeline
+        private static readonly TimeSpan FinishSerializationTimeout = TimeSpan.FromSeconds(3);
 
-        protected virtual void Cleanup() { }
+        private readonly Channel<ICounterPayload> _channel;
+        private readonly ChannelWriter<ICounterPayload> _channelWriter;
 
-        protected StreamingCounterLogger(Stream stream)
+        private ManualResetEvent _finishedSerialization = new(false);
+
+        protected abstract Task SerializeAsync(ICounterPayload counter);
+
+        protected StreamingCounterLogger()
         {
-            _stream = stream;
+            _channel = Channel.CreateBounded<ICounterPayload>(
+                new BoundedChannelOptions(BoundedChannelSize)
+                {
+                    AllowSynchronousContinuations = false,
+                    FullMode = BoundedChannelFullMode.DropWrite,
+                    SingleReader = true,
+                    SingleWriter = true
+                });
+            _channelWriter = _channel.Writer;
         }
 
         public void Log(ICounterPayload counter)
         {
-            //CONSIDER
-            //Ideally this would be an asynchronous api, but making this async would extend the lifetime of writing to the stream
-            //beyond the lifetime of the Counters pipeline.
-
-            SerializeCounter(_stream, counter);
+            _channelWriter.TryWrite(counter);
         }
 
         public void PipelineStarted()
         {
+            _ = ReadAndSerializeAsync();
         }
 
         public void PipelineStopped()
         {
-            Cleanup();
+            _channelWriter.Complete();
+
+            _finishedSerialization.WaitOne(FinishSerializationTimeout);
```

Comment:
This still seems like we're doing synchronous I/O: blocking an async worker thread until an IO operation completes (or times out).

Reply:
If we keep the Log call synchronous, but change PipelineStopped/PipelineStarted to async, we should be able to make this fully async.

Reply:
> This still seems like we're doing synchronous I/O: blocking an async worker thread until an IO operation completes (or times out).

There's no synchronous I/O in terms of writing to the target stream. Whether the worker thread blocks due to a synchronization primitive or not is orthogonal to I/O.

> If we keep the Log call synchronous, but change PipelineStopped/PipelineStarted to async, we should be able to make this fully async.

Sure, that would be nicer, and afford some minimal extra throughput on the thread calling `PipelineStarted` and `PipelineStopped`, but we'd still want to probably only await for some short-ish amount of time in case the reading/serialization loop is blocked for some reason; it's better to give up on waiting for completion (and probably log when that happens) instead of indefinitely blocking. I suppose that could be controlled by the caller via a `CancalletionToken` if we are able to plumb that through correctly.

Reply:
Per offline discussion, let's keep it this way and then change the PipelineFinished/PipelineStarted callbacks once we change the counter to support System.Diagnostics.Metrics